autoload/lsp/ui/vim/folding.vim

```diff
@@ -0,0 +1,178 @@
+let s:folding_ranges = {}
+let s:textprop_name = 'vim-lsp-folding-linenr'
+
+function! s:find_servers() abort
+    return filter(lsp#get_whitelisted_servers(), 'lsp#capabilities#has_folding_range_provider(v:val)')
+endfunction
+
+function! lsp#ui#vim#folding#fold(sync) abort
+    let l:servers = s:find_servers()
+
+    if len(l:servers) == 0
+        call lsp#utils#error('Folding not supported for ' . &filetype)
+        return
+    endif
+
+    let l:server = l:servers[0]
+    call lsp#ui#vim#folding#send_request(l:server, bufnr('%'), a:sync)
+endfunction
+
+function! s:set_textprops(buf) abort
+    " Use zero-width text properties to act as a sort of "mark" in the buffer.
+    " This is used to remember the line numbers at the time the request was
+    " sent. We will let Vim handle updating the line numbers when the user
+    " inserts or deletes text.
+
+    " Create text property, if not already defined
+    silent! call prop_type_add(s:textprop_name, {'bufnr': a:buf})
+
+    " First, clear all markers from the previous run
+    call prop_remove({'type': s:textprop_name, 'bufnr': a:buf})
+
+    " Add markers to each line
+    let l:i = 1
+    while l:i <= line('$')
+        call prop_add(l:i, 1, {'bufnr': a:buf, 'type': s:textprop_name, 'id': l:i})
+        let l:i += 1
+    endwhile
+endfunction
+
+function! lsp#ui#vim#folding#send_request(server_name, buf, sync) abort
+    if !lsp#capabilities#has_folding_range_provider(a:server_name)
+        return
+    endif
+
+    if has('textprop')
```

Comment:
Can we use `g:lsp_fold_enabled` to be consistent with other settings.

Reply:
Do you mean an additional check for `g:lsp_fold_enabled`:
```vim
if has('textprop') && g:lsp_fold_enabled
```
and then not send requests if `g:lsp_fold_enabled` is false?
Replacing the check for `has('textprop')` doesn't make sense, because folding works with or without textprops.

Reply:
I mean there are sometimes cases when servers are buggy. So even if folding works sometime we would want to disable folding completely.

Reply:
It can be enabled by default.